crates/runtime/src/libcalls.rs

```diff
@@ -435,49 +439,81 @@ unsafe fn memory_atomic_notify(
     vmctx: *mut VMContext,
     memory_index: u32,
     addr: u64,
-    _count: u32,
+    count: u32,
 ) -> Result<u32, TrapReason> {
     let memory = MemoryIndex::from_u32(memory_index);
-    let instance = (*vmctx).instance();
-    validate_atomic_addr(instance, memory, addr, 4, 4)?;
-    Err(
-        anyhow::anyhow!("unimplemented: wasm atomics (fn memory_atomic_notify) unsupported",)
-            .into(),
-    )
+    let instance = (*vmctx).instance_mut();
+    let spot = validate_atomic_addr(instance, memory, addr, 4, 4)?;
+
+    if count == 0 {
+        return Ok(0);
+    }
+
+    let unparked_threads = if count == u32::MAX {
+        spot.unpark_all(addr as _)
+    } else {
+        spot.unpark(addr as _, count as _)
+    };
+
+    Ok(unparked_threads as u32)
 }
 
 // Implementation of `memory.atomic.wait32` for locally defined memories.
 unsafe fn memory_atomic_wait32(
     vmctx: *mut VMContext,
     memory_index: u32,
     addr: u64,
-    _expected: u32,
-    _timeout: u64,
+    expected: u32,
+    timeout: u64,
 ) -> Result<u32, TrapReason> {
+    // convert to absolute timestamp as soon as possible
+    let timeout = (timeout as i64 >= 0).then(|| Duration::from_nanos(timeout));
+
     let memory = MemoryIndex::from_u32(memory_index);
-    let instance = (*vmctx).instance();
-    validate_atomic_addr(instance, memory, addr, 4, 4)?;
-    Err(
-        anyhow::anyhow!("unimplemented: wasm atomics (fn memory_atomic_wait32) unsupported",)
-            .into(),
-    )
+    let instance = (*vmctx).instance_mut();
+    let spot = validate_atomic_addr(instance, memory, addr, 4, 4)?;
+
+    // SAFETY: `addr` was validated by `validate_atomic_addr` above.
+    let atomic = unsafe { &*(addr as *const AtomicU32) };
+    // SAFETY: `addr` is a valid pointer into the given memory and unique to parking_lot_core.
+    match spot.park(
+        addr as usize,
+        || atomic.load(Ordering::SeqCst) == expected,
```

Comment:
Could you document that the `SeqCst` here matches the ordering specified by wasm itself?

Reply:
done